File: aoc2025/go/challenges/day1.go
package challenges

import (
	"log"
	"log/slog"
	"strconv"

	"aoc/utilities"

	logger "github.com/Bparsons0904/goLogger"
)

type DialInstruction struct {
	Direction string
	Step      int
}

type DialInstructions []DialInstruction

type Dial struct {
	Value int
	Next  *Dial
	Prev  *Dial
}

type DialList struct {
	Head *Dial
	Tail *Dial
}

func Day1() {
	log := logger.New("Day1")
	done := log.Timer("Day1 Timer")
	defer done()

	files := utilities.ReadFile("day1.part1")

	var dialInstructions DialInstructions
	dialInstructions.parseInstructions(files)

	var dialList DialList
	dialList.buildLinkedList()
	for {
		if dialList.Head.Value == 50 {
			break
		}
		dialList.next()
	}

	day1part1Count := 0
	day1part2Count := 0
	for _, dialInstruction := range dialInstructions {
		switch dialInstruction.Direction {
		case "R":
			for j := 0; j < dialInstruction.Step; j++ {
				dialList.next()
				if dialList.Head.Value == 0 {
					day1part2Count++
				}
			}

		case "L":
			for j := 0; j < dialInstruction.Step; j++ {
				dialList.prev()
				if dialList.Head.Value == 0 {
					day1part2Count++
				}
			}
		}
		if dialList.Head.Value == 0 {
			day1part1Count++
		}
	}

	slog.Info("day1", "part1", day1part1Count, "part2", day1part2Count)
}

func (d *DialList) next() {
	d.Head = d.Head.Next
}

func (d *DialList) prev() {
	d.Head = d.Head.Prev
}

func (d *DialList) buildLinkedList() {
	var prevNode *Dial
	for i := range 100 {
		node := &Dial{
			Value: i,
			Next:  nil,
			Prev:  prevNode,
		}

		if d.Head == nil {
			d.Head = node
		}

		if prevNode != nil {
			prevNode.Next = node
		}

		d.Tail = node
		prevNode = node
	}

	d.Tail.Next = d.Head
	d.Head.Prev = d.Tail
}

func (di *DialInstructions) parseInstructions(file []string) {
	for i := range file {
		var instruction DialInstruction
		instruction.Direction = string(file[i][0])
		step, err := strconv.Atoi(file[i][1:])
		if err != nil {
			log.Fatal(err)
		}
		instruction.Step = step

		*di = append(*di, instruction)
	}
}

func Day1_1() {
	log := logger.New("Day1_1")
	done := log.Timer("Timer")
	defer done()

	files := utilities.ReadFile("day1.part1")

	var dialInstructions DialInstructions
	dialInstructions.parseInstructions(files)

	step1Count := 0
	step2Count := 0
	currentValue := 50

	for _, dialInstruction := range dialInstructions {
		switch dialInstruction.Direction {
		case "R":
			newValue := currentValue + dialInstruction.Step
			if newValue >= 100 {
				if currentValue == 0 {
					step2Count += dialInstruction.Step / 100
				} else {
					step2Count += (dialInstruction.Step + currentValue) / 100
				}
			}
			currentValue = newValue % 100
		case "L":
			newValue := currentValue - dialInstruction.Step
			if currentValue == 0 {
				step2Count += dialInstruction.Step / 100
			} else if dialInstruction.Step >= currentValue {
				step2Count += (dialInstruction.Step-currentValue)/100 + 1
			}
			currentValue = ((newValue % 100) + 100) % 100
		}
		if currentValue == 0 {
			step1Count++
		}
	}

	slog.Info("day1", "part1", step1Count, "part2", step2Count)
}

File: aoc2025/zig/src/days/day1.zig
const std = @import("std");
const file_parser = @import("../file_parser.zig");
const print = std.debug.print;

const Dial = struct {
    value: u32,
    next: ?*Dial,
    prev: ?*Dial,
};

const DialList = struct {
    head: *Dial,
    tail: *Dial,
    
    fn init() !DialList {
        const allocator = std.heap.page_allocator;
        
        // Create 100 nodes in a circular linked list
        var nodes: [100]*Dial = undefined;
        for (0..100) |i| {
            const node = try allocator.create(Dial);
            node.value = @intCast(i);
            node.next = null;
            node.prev = null;
            nodes[i] = node;
        }
        
        // Link them together
        for (0..100) |i| {
            const next_idx = (i + 1) % 100;
            const prev_idx = if (i == 0) 99 else i - 1;
            
            nodes[i].next = nodes[next_idx];
            nodes[i].prev = nodes[prev_idx];
        }
        
        return DialList{
            .head = nodes[0],
            .tail = nodes[99],
        };
    }
    
    fn next(self: *DialList) void {
        self.head = self.head.next.?;
    }
    
    fn prev(self: *DialList) void {
        self.head = self.head.prev.?;
    }
};

const DialInstruction = struct {
    direction: u8,
    step: u32,
};

pub fn run() !void {
    try day1();
    try day1_1();
}

pub fn day1() !void {
    const allocator = std.heap.page_allocator;
    
    const files = try file_parser.readFile(allocator, "day1.part1");
    defer allocator.free(files);
    
    const dial_instructions = try allocator.alloc(DialInstruction, files.len);
    
    try parseInstructions(files, dial_instructions);
    
    var dial_list = try DialList.init();
    
    // Position at 50 like in the Go version
    while (dial_list.head.value != 50) {
        dial_list.next();
    }
    
    var day1part1Count: u32 = 0;
    var day1part2Count: u32 = 0;
    
    for (dial_instructions) |dial_instruction| {
        switch (dial_instruction.direction) {
            'R' => {
                for (0..dial_instruction.step) |_| {
                    dial_list.next();
                    if (dial_list.head.value == 0) {
                        day1part2Count += 1;
                    }
                }
            },
            'L' => {
                for (0..dial_instruction.step) |_| {
                    dial_list.prev();
                    if (dial_list.head.value == 0) {
                        day1part2Count += 1;
                    }
                }
            },
            else => return error.InvalidDirection,
        }
        
        if (dial_list.head.value == 0) {
            day1part1Count += 1;
        }
    }
    
    print("day1 part1: {d}, part2: {d}\n", .{ day1part1Count, day1part2Count });
}

pub fn day1_1() !void {
    const allocator = std.heap.page_allocator;
    
    const files = try file_parser.readFile(allocator, "day1.part1");
    defer allocator.free(files);
    
    const dial_instructions = try allocator.alloc(DialInstruction, files.len);
    
    try parseInstructions(files, dial_instructions);
    
    var step1Count: u32 = 0;
    var step2Count: u32 = 0;
    var currentValue: i32 = 50;
    
    for (dial_instructions) |dial_instruction| {
        switch (dial_instruction.direction) {
            'R' => {
                const newValue = currentValue + @as(i32, @intCast(dial_instruction.step));
                if (newValue >= 100) {
                    if (currentValue == 0) {
                        step2Count += @intCast(dial_instruction.step / 100);
                    } else {
                        step2Count += @intCast((dial_instruction.step + @as(u32, @intCast(currentValue))) / 100);
                    }
                }
                currentValue = @rem(newValue, 100);
            },
            'L' => {
                const newValue = currentValue - @as(i32, @intCast(dial_instruction.step));
                if (currentValue == 0) {
                    step2Count += @intCast(dial_instruction.step / 100);
                } else if (dial_instruction.step >= @as(u32, @intCast(currentValue))) {
                    step2Count += @intCast((dial_instruction.step - @as(u32, @intCast(currentValue))) / 100 + 1);
                }
                currentValue = @mod(newValue, 100);
            },
            else => return error.InvalidDirection,
        }
        
        if (currentValue == 0) {
            step1Count += 1;
        }
    }
    
    print("day1 part1 optimized: {d}, part2 optimized: {d}\n", .{ step1Count, step2Count });
}

fn parseInstructions(file_lines: []const []const u8, dial_instructions: []DialInstruction) !void {
    for (file_lines, 0..) |line, i| {
        if (line.len == 0) continue;
        
        dial_instructions[i] = DialInstruction{
            .direction = line[0],
            .step = try std.fmt.parseInt(u32, line[1..], 10),
        };
    }
}